languages/csharp/exercises/concept/nullability/.docs/instructions.md

```diff
@@ -0,0 +1,30 @@
+In this exercise you'll be writing code to helping printing name
+badges for factory employees.
+
+Employees have an id, a name and a department name. Badge labels are
+formatted as follows: `"[id] - [name] - [DEPARTMENT]"`
+
+### 1. Generate the badge label from employee properties
+
+Writing a static method called `Badge.Label()` to compute the string
+to be printed on the badge:
+
+```csharp
+Badge.Label(734, "Ernest Johnny Payne", "Strategic Communication");
+// => "[734] - Ernest Johnny Payne - STRATEGIC COMMUNICATION"
+```
+
+The department name is optional. If someone has no department, just
+print `GUEST` as their department name.
+
+### 2. Word wrap the label text so it fits on a badge
+
+Writing a static method called `Badge.PrintLabel()`:
+
+```csharp
+Badge.PrintLabel("[734] - Ernest Johnny Payne - STRATEGIC COMMUNICATION", 30);
+// => "[734] - Ernest Johnny Payne - \nSTRATEGIC COMMUNICATION"
+```
+
+The width is optional, so if it is not provided, just return the
```

Comment:
Could you make this a separate task?

Reply:
Do you mean  adding a third method? 

It's weird to split a method into two tasks since if it is a method it does one thing, not two. I would prefer changing the wording to imply that the width is optional (just like on the first task). 

**PS1:** To me it's weird to have too many "tasks" on an exercise, since it's something that only exists on the docs... at least that's not how I am used to work on exercises (maybe that's a difference between v2 and v3 that I am not aware of), but when doing v2 exercises, I usually read the spec, then work on the tests, one by one.... 

**PS2:** Sometimes I think It'd better to do something more than just returning the string as is when no `maxWidth` is given... for example, we could split the string fields and then make sure it still fits on the given width if any. It's just that if the method does nothing if no width is given, there's no point in calling the method.

```
[734]
Ernest Johnny Payne
STRATEGIC COMMU
NICATION
```

Another idea, we could simply put it in a simple HTML document...

```
<html>
<body>
<em>
[734] - Ernest Johnny Payne
STRATEGIC COMMUNICATION
</em>
</body>
</html>
```

Reply:
> To me it's weird to have too many "tasks" on an exercise, since it's something that only exists on the docs... at least that's not how I am used to work on exercises (maybe that's a difference between v2 and v3 that I am not aware of), but when doing v2 exercises, I usually read the spec, then work on the tests, one by one....

This is indeed something that is going to be very different in v3, as we'll have in-browser coding support, which we'll expect will be quite popular. The in-browser coding experience won't show the tests to the student. It will show test results (assertion messages), but it won't show the actual test code. Splitting up things into different tasks can therefore help students focus on specific bits of the implementation. Does that help explain why I prefer splitting things up?

Reply:
> Sometimes I think It'd better to do something more than just returning the string as is when no maxWidth is given... for example, we could split the string fields and then make sure it still fits on the given width if any. It's just that if the method does nothing if no width is given, there's no point in calling the method.

I do agree with this, although I'm not sure I like the suggested functionality better. I don't think it makes sense to have one method return a label string without newlines and then later on split it again.

Reply:
> Does that help explain why I prefer splitting things up?

ok, I will add another task, but I assume I do not need to split the method,

Reply:
> I do agree with this, although I'm not sure I like the suggested functionality better. I don't think it makes sense to have one method return a label string without newlines and then later on split it again.

Yes, I am not happy either with my suggestion... I will try to find something better... I just want to avoid the 

```csharp
if (maxWidth == null) return label;
```

Reply:
> ok, I will add another task, but I assume I do not need to split the method,

Correct.

> Yes, I am not happy either with my suggestion... I will try to find something better... I just want to avoid the

Agreed. What if we changed things around to make the ID optional instead of the max width?

```csharp
string Label(int id?, string name, string? department)
{
    if (id == null)
       return $"{name} - {department ?? "GUEST"}";

    return $"[{id}] - {name} - {department ?? "GUEST"}";
}
```

And then the `PrintLabel` method could always take the max width as a regular, non-nullable int.

Maybe something like this could work? It would mean you get the keep the max width limiting functionality, which _does_ make a lot of sense from a story viewpoint and we get rid of the `if (label == null) return label` bit.

Reply:
Yes I like it :)
Then I will need to change the tasks too, I'd prefer something like:

1. Implement `Label` as if parameters are never `null`
2. Deal with the `null` `id` and `department`
3. Implement `PrintLabel`

It will have an increasing level of difficulty, which  will be better than the 3 tasks we have right now.

Reply:
Yes! I like what you've suggested! Sounds great. Sorry for the churn on this, but nullability _is_ a difficulty thing to get right. Which is of course also why they added the nullablility features to C# 8 :)

Reply:
Hi, just pushed a new version... the problem now is: the method `PrintLabel` has nothing to do with nullability anymore...

What about adding nullable prefix and a suffix that the student needs to concatenate to the string (if present) before wrapping?

Reply:
@ErikSchierboom

Reply:
Just pushed a comment implementing that. Just added an optional `prefix` to be added before print.